src/util/util_fps_limiter.cpp

```diff
@@ -0,0 +1,159 @@
+#include <thread>
+
+#include "thread.h"
+#include "util_fps_limiter.h"
+#include "util_string.h"
+
+#include "./log/log.h"
+
+namespace dxvk {
+  
+  FpsLimiter::FpsLimiter() {
+
+  }
+
+
+  FpsLimiter::~FpsLimiter() {
+
+  }
+
+
+  void FpsLimiter::setTargetFrameRate(double frameRate) {
+    std::lock_guard<std::mutex> lock(m_mutex);
+
+    m_targetInterval = frameRate > 0.0
+      ? NtTimerDuration(int64_t(double(NtTimerDuration::period::den) / frameRate))
+      : NtTimerDuration::zero();
+
+    if (isEnabled() && !m_initialized)
+      initialize();
+  }
+
+
+  void FpsLimiter::setDisplayRefreshRate(double refreshRate) {
+    std::lock_guard<std::mutex> lock(m_mutex);
+
+    m_refreshInterval = refreshRate > 0.0
+      ? NtTimerDuration(int64_t(double(NtTimerDuration::period::den) / refreshRate))
+      : NtTimerDuration::zero();
+  }
+
+
+  void FpsLimiter::delay(bool vsyncEnabled) {
+    std::lock_guard<std::mutex> lock(m_mutex);
```

Comment:
Why does this need a mutex?

setTargetFrameRate should only happen at startup, right?

Reply:
I guess also on swapchain mode change but we aren't presenting then.

Reply:
We can have a present queued up or executing on the queue thread while the app thread performs a mode change and calls `setDisplayRefreshRate`.

This does mean that not every frame during mode chages is necessarily perfect, but I'd rather do this than make presentation itself more complicated than it already is.